src/system/SystemObject.h

```diff
@@ -228,10 +261,18 @@ template <class T, unsigned int N>
 inline T * ObjectPool<T, N>::TryCreate(Layer & aLayer)
 {
     T * lReturn = nullptr;
-    unsigned int lIndex;
-#if CHIP_SYSTEM_CONFIG_PROVIDE_STATISTICS
-    unsigned int lNumInUse = 0;
-#endif
+
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+    mtx.lock();
+    mObjects.push_back(std::unique_ptr<T>(new T()));
```

Comment:
We should be constructing before we acquire the lock.  This isn't in the critical path.

I think we likely should also not be calling TryCreate with the lock held.  Isn't this just happening as a convenience to get the `.back()` pointer with the lock held?  Get and and store that pointer, release the lock an then call TryCreate.

Reply:
without this mutex protection, we are not even pass multi-thread unit test

Reply:
msandstedt@ is just suggesting to do something along the lines of

```
auto ptr = std::make_unique<T>();
if (!ptr->TryCreate(aLayer, sizeof(T))
  return nullptr;

std::lock_guard<std::mutex> lock(mMutex);
mObjects.push_back(std::move(ptr));
return mObjects.back().get();
```

Please use `std::lock_guard`.

Reply:
Correct.  The critical path is the code that touches `mObjects`.  I'm suggesting narrowing the scope of the lock to just the path that needs to operate on `mObjects`.